### ISSUE TITLE
Asynchronous Request-Response

### ISSUE THREAD

**Issue by ishbir** [nats-io/cnats#28]
I'm using cnats in an application which requires asynchronous responses to a request. I know that `natsConnection_Request` exists but it blocks until the response has been returned from the broker.

I reasoned that asynchronicity could be achieved using `natsConnection_Subscribe` to a temporary subject and calling `natsSubscription_AutoUnsubscribe(sub, 1)` to make sure that it's unsubscribed right after. However, the problem is that the subscription also needs to be destroyed and the memory freed. How could this be done without modifying cnats (if at all)?

**Comment by kozlovic** [comment]
You can indeed use `natsConnection_PublishRequest` (or its variant `natsConnection_PublishRequestString`) to publish a request and use any subscription API call (such as `natsConnection_Subscribe`) to subscribe to the reply subject.

As for the memory cleanup, since you use auto-unsubscribe, you are guaranteed that your message handler will not receive more than 1 message for this subscription. It is then conceivable to destroy the subscription (call `natsSubscription_Destroy`) from the message handler. Note that `natsMsgHandler` has a `natsSubscription` in the definition, so you have the subscription to destroy handed to you.
 
If you use the same message handler code with different subscriptions having different auto-unsubscribe values (or no auto-unsubscribe at all), then you may need to use the closure to pass some info on when to call the destroy.

**Comment by ishbir** [comment]
An edge case: what happens when no response is received (either the broker or the server responding is at fault)? `natsSubscription` would stay in memory forever. And if an application sends many requests per second, it could lead to enormous memory leaks and crashes. Any idea how this could be solved?

I thought of having some sort of a timeout, like the synchronous version. But again, this requires access to the internal threadpool to implement effectively as I don't want to create my own threadpool just for avoiding this scenario.

**Comment by kozlovic** [comment]
If I understand correctly (and correct me if I am wrong):
1. You create an Inbox
2. You setup a subscription (with auto-unsubscribe 1) on that Inbox
3. You call `natsConnection_PublishRequest` on the request subject and the inbox as the reply.
If that's correct, what I suggested is that you call `natsSubscription_Destroy` in the message handler when processing the reply.
To solve the "no response" issue, you could have a timer of a value corresponding to your desired timeout and pass the inbox created in step 1. When the timer fires, you should send a special message to the inbox. Your message handler should recognize those messages as a "timeout" message and simply destroy the subscription without applying the business logic.
Would that solve your problem?

**Comment by ishbir** [comment]
I guess that could work if no reply is received but the connection remains alive. But if the connection is lost waiting on the reply, then sending a message to the inbox will also fail (not too sure about this though) and the resources will never be freed.

Another problem with this approach is that it requires that there be a separate thread in my application that takes care of these timeouts. I'm not very familiar with the internal architecture of cnats but if there's a way to do it such that no new threads/threadpools are created, that would be much better.

**Comment by kozlovic** [comment]
About lost connection, when that happens, assuming that the connection cannot reconnect, the publish will fail. You should then release the resources. I guess that you would need to do some bookkeeping indeed.

Now, to step back: when using `natsConnection_Request`, this call creates the subscription and releases it, but this is a synchronous subscription, that is, no extra thread is created. When using an asynchronous subscription, the library will create a thread responsible to deliver messages. If you use the subscription for a single reply, this may not be that efficient (lots of threads created/destroyed).

You may want to look at alternatives such as using a single subscription and deal with possible multiple replies for a single request, or use a synchronous subscription and call `natsSubscription_NextMsg` when you want (either from a different thread or after executing some code passed the publish request call).

If you could describe a bit more your use case, I may be able to propose better approach than the one above?

**Comment by ishbir** [comment]
What I'm trying to achieve is an RPC style interface between a client and a 'server' wherein the server replies to client's requests using NATS and may even stream its own data. The issue is that my C++11 application is completely async and extensively uses callbacks in its function calls for response that could be delayed.

Using a single subscription and dealing with possible multiple replies for a single request doesn't seem all that very different from the previously mentioned approach (except that it doesn't auto-unsubscribe after the first message) or maybe I misunderstood what you're trying to say. Using `natsSubscription_NextMsg` on a synchronous subscription means having to implement a thick glue (using something like Boost ASIO maybe) between the callbacks of the application and the synchronous layer. This could be done but I'd like to keep it as simple and thus, maintainable, as possible.

**Comment by kozlovic** [comment]
Ok, so to sump up, the issue is that we need to find a way to destroy the asynchronous subscription. The easy way is when a reply is received, and since you use auto-unsubscribe(1), simply call `natsSubscription_Destroy` in the message handler. However, this does not cover the case when no reply is received.

You are not keen on having your application create threads (which would be a solution: create a thread which would simply call `natsSubscription_NextMsg()` which allows for a timeout. After the reply is received (or timeout) destroy the subscription before the thread exit).

Note that if cnats provide a new API to achieve this (or modify the current async delivery thread to wait with a timeout), the number of created/destroyed threads will be the same, it's just that you - the developer - will not have to manage them.

That being said, let's think of new APIs that would solve your problem. What would you need exactly? I am thinking that you would need something to the effect of creating an asynchronous subscription that auto-destroys and allows a timeout. For instance:
```
natsStatus
natsConnection_SubscribeTimeout(
    natsSubscription **sub, 
    natsConnection *nc,
    const char *subject,
    int64_t timeout,
    natsMsgHandler cb,
    void *cbClosure)
```
We could say that after this call, a delivery thread is started but instead of waiting only for the subscription to be explicitly closed or receiving the max number of messages (auto-unsubscribe), it also waits for the given timeout. When the delivery thread exits (either after receiving the max number of messages, or timing-out), the subscription is freed. This would allow the user to setup the subscription without having to explicitly call `natsSubscription_Destroy` and still make sure that the subscription does not linger past the given timeout.
With this approach, and since we directly use pointers, there is no way to allow the call of `natsSubscription_Destroy` since there could be a race where the delivery thread is about to exit and takes responsibility of freeing the subscription while another thread tries to explicitly destroy it and access a freed pointer. In other words, once the subscription would be started this way, there would be no safe way to call ANY subscription APIs. So even auto-unsubscribe call is potentially unsafe (assume a very small timeout and a delayed call to auto-unsubscribe). This means that the auto-unsubscribe value should probably be passed to the create call...

**Comment by ishbir** [comment]
> You are not keen on having your application create threads (which would be a solution: create a thread which would simply call natsSubscription_NextMsg() which allows for a timeout. After the reply is received (or timeout) destroy the subscription before the thread exit).

This means one thread per reply which has an unreasonably high cost. Even if there were a way to use subscriptions asynchronously in the way that I wanted, this cost remains constant. One thread per subscription is a model that simply cannot scale to where we want it to (maybe hundreds of requests per second). I wish I had known this before we went with cnats.

Another idea that had occurred to me was to use a fixed reply channel for subscriptions of a particular type and embed a `uuid` or something of the sort within the messages received, then have a dispatcher route the messages. However, this is extremely complex. It would be easier to just re-write cnats to make it 'truly' async and not use blocking IO on threads, which is a bad idea for high performance applications.

My team and I are just going ahead with blocking requests for now; just because it's already there and to have a demo ASAP. Maybe in the future we'll think of a solution to this problem (contribute code to cnats to make it truly async) or switch to ZeroMQ.

**Comment by kozlovic** [comment]
Yes. Synchronous request/reply do not create any thread, but an asynchronous subscription requires a thread to do the delivery.

Depending on your use case, yes, you may be better off having a single subscription, as I described earlier, and use some info in the message to correlate with the request.

The connection has a thread to receive data from the socket. When processing a message, it is right now inserting that message into the subscription's list that is dispatched in the subscription's thread. You could modify cnats to invoke the subscriber's callback directly from the connection's process message function (executing in the readLoop). But be careful, this is going to block the connection, so it may be a problem if something in the message handler tries to communicate through that same connection.

Otherwise, we could look into a model where a single thread (per connection) invokes the message handlers for all subscriptions on that connection (instead of having a thread per subscription). Here again, you may experience delay in message handlers since their execution will be serialized (even for different subscriptions).

At any rate, we value your feedback and are looking forward to your contributions!

**Comment by ishbir** [comment]
> Otherwise, we could look into a model where a single thread (per connection) invokes the message handlers for all subscriptions on that connection (instead of having a thread per subscription). Here again, you may experience delay in message handlers since their execution will be serialized (even for different subscriptions).

What if there were one thread per connection (like it is right now; although again, not a great idea) and a task queue-cum-thread pool with say 4 threads that call all the subscriber's callback functions. That way even if there are 10000 subscriptions, there'll always be `4+n` threads where `n` is the number of connections. IMO, this means minimal changes for maximum results.

However, this still doesn't solve my original problem of resource management. Will need to think more about it.

**Comment by kozlovic** [comment]
You need to have a thread doing socket receive. If it's not in cnats, it has to be somewhere! Since `1.3` we allow you to plug with an event loop (so far we have adapters for `libuv` and `libevent`).
Yes, like I said, we could have a thread (or pool of threads) per connection: note that with the pool, you would have to bind a given subscription to one of these threads because users expect messages from a given subscription to be delivered in order. If you keep adding all these, you may start to have a degradation for normal cases where users don't have thousands of subscriptions per connection, so this would have to be a configuration enabled behavior in my opinion.

**Comment by ishbir** [comment]
> You need to have a thread doing socket receive. If it's not in cnats, it has to be somewhere! Since `1.3` we allow you to plug with an event loop (so far we have adapters for `libuv` and `libevent`).

Ah, of course, event loops are supported. Slipped my mind.

> Yes, like I said, we could have a thread (or pool of threads) per connection: note that with the pool, you would have to bind a given subscription to one of these threads because users expect messages from a given subscription to be delivered in order.

Guaranteed in-order delivery does complicate matters. I can't think of any solution to this right now but this is one of the things that makes you wish that a good implementation of something like go-routines existed for C/C++. :)

> If you keep adding all these, you may start to have a degradation for normal cases where users don't have thousands of subscriptions per connection, so this would have to be a configuration enabled behavior in my opinion.

I'm gonna have to disagree here. Any non-trivial application is going to have more then 3-4 subscriptions. Even if they aren't a thousand, having a dozen threads for a dozen subscriptions isn't justified. Of course the size of the thread pool could be configurable, maybe `2` could be a reasonable default.

I'll keep you updated if I have any ideas.

**Comment by kozlovic** [comment]
> Guaranteed in-order delivery does complicate matters. I can't think of any solution to this right now but this is one of the things that makes you wish that a good implementation of something like go-routines existed for C/C++. :)

There is actually: http://libmill.org/

> I'm gonna have to disagree here. Any non-trivial application is going to have more then 3-4 subscriptions. Even if they aren't a thousand, having a dozen threads for a dozen subscriptions isn't justified.

You could have an application listening only to a given subject, say a specific quote, or some orders, etc.. What I meant by the possible degradation is that having that in-between thread will need some locking and subscriptions mapping that may be detrimental (locking under contention) for a smaller number of subscriptions. But don't get me wrong, I get your point: with a certain number of subscriptions, the number of threads will simply not work (limit on number of threads in a process, resources, etc...).

> I'll keep you updated if I have any ideas.

Please do! And as soon as I have a bit of time, I will start looking at what I could do there myself. Thanks again.

**Comment by kozlovic** [comment]
@ishbir I have made some progress on this and will probably create a branch soon for you to have a look if you are still interested. This branch would only contain the thread pool responsible for delivering messages for now (not the async subscribe with timeout, for which I have an idea how to solve).

I originally went the route of having a thread pool per connection. The advantage was that there would be a need for a single new API, say `natsOptions_SetMsgDeliveryThreadPool(int size)`. If by default the thread pool is 0, the library would behave as it is today. If you call the above with a size > 0, then subscribers created from the connection using this `natsOptions` object would then not create their own delivery thread but use the connection's one.

But re-reading your comments, it sounded that it should be even more global (a library level pool). So I changed to this, but we have now the need for 2 different APIs. One for instructing the subscribers to use the global thread pool (say something like `natsOptions_UseGlobalMsgDelivery(bool global)`) and one to optionally change the size of the pool, at the library level: `nats_SetMsgDeliveryPoolSize(int max)`. There would be a default of say 2 workers (that would be lazy created by the way, so thread created only when at least 2 subscribers are created), but one could increase the pool size with this call.

Which one do you think is better?

**Comment by kozlovic** [comment]
Here is the branch I mentioned in the previous post: https://github.com/nats-io/cnats/tree/global_msgs_delivery

Specifically, here is the main commit: https://github.com/nats-io/cnats/commit/f0a19389fbc7cb7a3247af82cecbb128f6f37e5f

**Comment by ishbir** [comment]
@kozlovic I'm more inclined to go with the library level threadpool just because it's more flexible. Give me some time to review and test the code and I'll get back to you.

**Comment by kozlovic** [comment]
@ishbir Thanks! I have fixed an issue on that branch, the workers load-balancing was not working well ;-)

**Comment by kozlovic** [comment]
@ishbir I have deleted the branch and created a new one, along with a PR: https://github.com/nats-io/cnats/pull/34. I believe this addresses all the requirements that you had: scalability with a central message delivery thread pool, and a way to have asynchronous subscriptions that timeout.

I have solved the issue with auto-destroy of the subscription that could have been dangerous, but invoking the message handler with a NULL message when the subscription times-out. The user can then destroy safely the subscription then.

Check https://github.com/nats-io/cnats/blob/7ad911abe505a1c88250bf565259c7df2efb3711/examples/asynctimeout.c for an example on how to use the new async subscriptions with timeout APIs.
